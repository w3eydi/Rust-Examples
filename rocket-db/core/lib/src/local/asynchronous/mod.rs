//! Asynchronous local dispatching of requests.
//!
//! This module contains the `asynchronous` variant of the `local` API: it can
//! be used with `#[rocket::async_test]` or another asynchronous test harness.
//! For the blocking variant, see [`blocking`](super::blocking).
//!
//! See the [top-level documentation](super) for more usage details.

mod client;
mod request;
mod response;

pub use client::*;
pub use request::*;
pub use response::*;
