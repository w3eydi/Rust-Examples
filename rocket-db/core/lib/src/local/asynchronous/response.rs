use std::io;
use std::future::Future;
use std::{pin::Pin, task::{Context, Poll}};

use tokio::io::{AsyncRead, ReadBuf};

use crate::http::CookieJar;
use crate::{Request, Response};

/// An `async` response from a dispatched [`LocalRequest`](super::LocalRequest).
///
/// This `LocalResponse` implements [`tokio::io::AsyncRead`]. As such, if
/// [`into_string()`](LocalResponse::into_string()) and
/// [`into_bytes()`](LocalResponse::into_bytes()) do not suffice, the response's
/// body can be read directly:
///
/// ```rust
/// # #[macro_use] extern crate rocket;
/// use std::io;
///
/// use rocket::local::asynchronous::Client;
/// use rocket::tokio::io::AsyncReadExt;
/// use rocket::http::Status;
///
/// #[get("/")]
/// fn hello_world() -> &'static str {
///     "Hello, world!"
/// }
///
/// #[launch]
/// fn rocket() -> _ {
///     rocket::build().mount("/", routes![hello_world])
///     #    .configure(rocket::Config::debug_default())
/// }
///
/// # async fn read_body_manually() -> io::Result<()> {
/// // Dispatch a `GET /` request.
/// let client = Client::tracked(rocket()).await.expect("valid rocket");
/// let mut response = client.get("/").dispatch().await;
///
/// // Check metadata validity.
/// assert_eq!(response.status(), Status::Ok);
/// assert_eq!(response.body().preset_size(), Some(13));
///
/// // Read 10 bytes of the body. Note: in reality, we'd use `into_string()`.
/// let mut buffer = [0; 10];
/// response.read(&mut buffer).await?;
/// assert_eq!(buffer, "Hello, wor".as_bytes());
/// # Ok(())
/// # }
/// # rocket::async_test(read_body_manually()).expect("read okay");
/// ```
///
/// For more, see [the top-level documentation](../index.html#localresponse).
pub struct LocalResponse<'c> {
    _request: Box<Request<'c>>,
    response: Response<'c>,
    cookies: CookieJar<'c>,
}

impl<'c> LocalResponse<'c> {
    pub(crate) fn new<F, O>(req: Request<'c>, f: F) -> impl Future<Output = LocalResponse<'c>>
        where F: FnOnce(&'c Request<'c>) -> O + Send,
              O: Future<Output = Response<'c>> + Send
    {
        // `LocalResponse` is a self-referential structure. In particular,
        // `inner` can refer to `_request` and its contents. As such, we must
        //   1) Ensure `Request` has a stable address.
        //
        //      This is done by `Box`ing the `Request`, using only the stable
        //      address thereafter.
        //
        //   2) Ensure no refs to `Request` or its contents leak with a lifetime
        //      extending beyond that of `&self`.
        //
        //      We have no methods that return an `&Request`. However, we must
        //      also ensure that `Response` doesn't leak any such references. To
        //      do so, we don't expose the `Response` directly in any way;
        //      otherwise, methods like `.headers()` could, in conjunction with
        //      particular crafted `Responder`s, potentially be used to obtain a
        //      reference to contents of `Request`. All methods, instead, return
        //      references bounded by `self`. This is easily verified by noting
        //      that 1) `LocalResponse` fields are private, and 2) all `impl`s
        //      of `LocalResponse` aside from this method abstract the lifetime
        //      away as `'_`, ensuring it is not used for any output value.
        let boxed_req = Box::new(req);
        let request: &'c Request<'c> = unsafe { &*(&*boxed_req as *const _) };

        async move {
            let response: Response<'c> = f(request).await;
            let mut cookies = CookieJar::new(request.rocket().config());
            for cookie in response.cookies() {
                cookies.add_original(cookie.into_owned());
            }

            LocalResponse { cookies, _request: boxed_req, response, }
        }
    }
}

impl LocalResponse<'_> {
    pub(crate) fn _response(&self) -> &Response<'_> {
        &self.response
    }

    pub(crate) fn _cookies(&self) -> &CookieJar<'_> {
        &self.cookies
    }

    pub(crate) async fn _into_string(mut self) -> io::Result<String> {
        self.response.body_mut().to_string().await
    }

    pub(crate) async fn _into_bytes(mut self) -> io::Result<Vec<u8>> {
        self.response.body_mut().to_bytes().await
    }

    #[cfg(feature = "json")]
    async fn _into_json<T: Send + 'static>(self) -> Option<T>
        where T: serde::de::DeserializeOwned
    {
        self.blocking_read(|r| serde_json::from_reader(r)).await?.ok()
    }

    #[cfg(feature = "msgpack")]
    async fn _into_msgpack<T: Send + 'static>(self) -> Option<T>
        where T: serde::de::DeserializeOwned
    {
        self.blocking_read(|r| rmp_serde::from_read(r)).await?.ok()
    }

    #[cfg(any(feature = "json", feature = "msgpack"))]
    async fn blocking_read<T, F>(mut self, f: F) -> Option<T>
        where T: Send + 'static,
              F: FnOnce(&mut dyn io::Read) -> T + Send + 'static
    {
        use tokio::sync::mpsc;
        use tokio::io::AsyncReadExt;

        struct ChanReader {
            last: Option<io::Cursor<Vec<u8>>>,
            rx: mpsc::Receiver<io::Result<Vec<u8>>>,
        }

        impl std::io::Read for ChanReader {
            fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
                loop {
                    if let Some(ref mut cursor) = self.last {
                        if cursor.position() < cursor.get_ref().len() as u64 {
                            return std::io::Read::read(cursor, buf);
                        }
                    }

                    if let Some(buf) = self.rx.blocking_recv() {
                        self.last = Some(io::Cursor::new(buf?));
                    } else {
                        return Ok(0);
                    }
                }
            }
        }

        let (tx, rx) = mpsc::channel(2);
        let reader = tokio::task::spawn_blocking(move || {
            let mut reader = ChanReader { last: None, rx };
            f(&mut reader)
        });

        loop {
            // TODO: Try to fill as much as the buffer before send it off?
            let mut buf = Vec::with_capacity(1024);
            match self.read_buf(&mut buf).await {
                Ok(n) if n == 0 => break,
                Ok(_) => tx.send(Ok(buf)).await.ok()?,
                Err(e) => {
                    tx.send(Err(e)).await.ok()?;
                    break;
                }
            }
        }

        // NOTE: We _must_ drop tx now to prevent a deadlock!
        drop(tx);

        reader.await.ok()
    }

    // Generates the public API methods, which call the private methods above.
    pub_response_impl!("# use rocket::local::asynchronous::Client;\n\
        use rocket::local::asynchronous::LocalResponse;" async await);
}

impl AsyncRead for LocalResponse<'_> {
    fn poll_read(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &mut ReadBuf<'_>,
    ) -> Poll<io::Result<()>> {
        Pin::new(self.response.body_mut()).poll_read(cx, buf)
    }
}

impl std::fmt::Debug for LocalResponse<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self._response().fmt(f)
    }
}
