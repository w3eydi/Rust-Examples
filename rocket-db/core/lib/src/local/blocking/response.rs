use std::io;
use tokio::io::AsyncReadExt;

use crate::{Response, local::asynchronous, http::CookieJar};

use super::Client;

/// A `blocking` response from a dispatched [`LocalRequest`](super::LocalRequest).
///
/// This `LocalResponse` implements [`io::Read`]. As such, if
/// [`into_string()`](LocalResponse::into_string()) and
/// [`into_bytes()`](LocalResponse::into_bytes()) do not suffice, the response's
/// body can be read directly:
///
/// ```rust
/// # #[macro_use] extern crate rocket;
/// use std::io::{self, Read};
///
/// use rocket::local::blocking::Client;
/// use rocket::http::Status;
///
/// #[get("/")]
/// fn hello_world() -> &'static str {
///     "Hello, world!"
/// }
///
/// #[launch]
/// fn rocket() -> _ {
///     rocket::build().mount("/", routes![hello_world])
///     #    .configure(rocket::Config::debug_default())
/// }
///
/// # fn read_body_manually() -> io::Result<()> {
/// // Dispatch a `GET /` request.
/// let client = Client::tracked(rocket()).expect("valid rocket");
/// let mut response = client.get("/").dispatch();
///
/// // Check metadata validity.
/// assert_eq!(response.status(), Status::Ok);
/// assert_eq!(response.body().preset_size(), Some(13));
///
/// // Read 10 bytes of the body. Note: in reality, we'd use `into_string()`.
/// let mut buffer = [0; 10];
/// response.read(&mut buffer)?;
/// assert_eq!(buffer, "Hello, wor".as_bytes());
/// # Ok(())
/// # }
/// # read_body_manually().expect("read okay");
/// ```
///
/// For more, see [the top-level documentation](../index.html#localresponse).
pub struct LocalResponse<'c> {
    pub(in super) inner: asynchronous::LocalResponse<'c>,
    pub(in super) client: &'c Client,
}

impl LocalResponse<'_> {
    fn _response(&self) -> &Response<'_> {
        &self.inner._response()
    }

    pub(crate) fn _cookies(&self) -> &CookieJar<'_> {
        self.inner._cookies()
    }

    fn _into_string(self) -> io::Result<String> {
        self.client.block_on(self.inner._into_string())
    }

    fn _into_bytes(self) -> io::Result<Vec<u8>> {
        self.client.block_on(self.inner._into_bytes())
    }

    #[cfg(feature = "json")]
    fn _into_json<T: Send + 'static>(self) -> Option<T>
        where T: serde::de::DeserializeOwned
    {
        serde_json::from_reader(self).ok()
    }

    #[cfg(feature = "msgpack")]
    fn _into_msgpack<T: Send + 'static>(self) -> Option<T>
        where T: serde::de::DeserializeOwned
    {
        rmp_serde::from_read(self).ok()
    }

    // Generates the public API methods, which call the private methods above.
    pub_response_impl!("# use rocket::local::blocking::Client;\n\
        use rocket::local::blocking::LocalResponse;");
}

impl io::Read for LocalResponse<'_> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        self.client.block_on(self.inner.read(buf))
    }
}

impl std::fmt::Debug for LocalResponse<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self._response().fmt(f)
    }
}
