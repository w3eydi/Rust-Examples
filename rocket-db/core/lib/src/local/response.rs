macro_rules! getter_method {
    ($doc_prelude:literal, $desc:literal, $f:ident -> $r:ty) => (
        getter_method!(@$doc_prelude, $f, $desc, $r,
            concat!("let ", stringify!($f), " = response.", stringify!($f), "();"));
    );
    (@$doc_prelude:literal, $f:ident, $desc:expr, $r:ty, $use_it:expr) => (
        /// Returns the
        #[doc = $desc]
        /// of `self`.
        ///
        /// # Example
        ///
        /// ```rust
        #[doc = $doc_prelude]
        ///
        /// # Client::_test(|_, _, response| {
        /// let response: LocalResponse = response;
        #[doc = $use_it]
        /// # });
        /// ```
        #[inline(always)]
        pub fn $f(&self) -> $r {
            self._response().$f()
        }
    )
}

macro_rules! pub_response_impl {
    ($doc_prelude:literal $($prefix:tt $suffix:tt)?) =>
{
    getter_method!($doc_prelude, "HTTP status",
        status -> crate::http::Status);

    getter_method!($doc_prelude, "Content-Type, if a valid one is set,",
        content_type -> Option<crate::http::ContentType>);

    getter_method!($doc_prelude, "HTTP headers",
        headers -> &crate::http::HeaderMap<'_>);

    /// Return a cookie jar containing the HTTP cookies in the response.
    ///
    /// # Example
    ///
    /// ```rust
    #[doc = $doc_prelude]
    ///
    /// # Client::_test(|_, _, response| {
    /// let response: LocalResponse = response;
    /// let string = response.cookies();
    /// # });
    /// ```
    #[inline(always)]
    pub fn cookies(&self) -> &crate::http::CookieJar<'_> {
        self._cookies()
    }

    getter_method!($doc_prelude, "response body, if there is one,",
        body -> &crate::response::Body<'_>);

    /// Consumes `self` and reads the entirety of its body into a string.
    ///
    /// If reading fails, the body contains invalid UTF-8 characters, or the
    /// body is unset in the response, returns `None`. Otherwise, returns
    /// `Some`. The string may be empty if the body is empty.
    ///
    /// # Example
    ///
    /// ```rust
    #[doc = $doc_prelude]
    ///
    /// # Client::_test(|_, _, response| {
    /// let response: LocalResponse = response;
    /// let string = response.into_string();
    /// # });
    /// ```
    #[inline(always)]
    pub $($prefix)? fn into_string(self) -> Option<String> {
        if self._response().body().is_none() {
            return None;
        }

        self._into_string() $(.$suffix)? .ok()
    }

    /// Consumes `self` and reads the entirety of its body into a `Vec` of
    /// bytes.
    ///
    /// If reading fails or the body is unset in the response, returns `None`.
    /// Otherwise, returns `Some`. The returned vector may be empty if the body
    /// is empty.
    ///
    /// # Example
    ///
    /// ```rust
    #[doc = $doc_prelude]
    ///
    /// # Client::_test(|_, _, response| {
    /// let response: LocalResponse = response;
    /// let bytes = response.into_bytes();
    /// # });
    /// ```
    #[inline(always)]
    pub $($prefix)? fn into_bytes(self) -> Option<Vec<u8>> {
        if self._response().body().is_none() {
            return None;
        }

        self._into_bytes() $(.$suffix)? .ok()
    }

    /// Consumes `self` and deserializes its body as JSON without buffering in
    /// memory.
    ///
    /// If deserialization fails or the body is unset in the response, returns
    /// `None`. Otherwise, returns `Some`.
    ///
    /// # Example
    ///
    /// ```rust
    #[doc = $doc_prelude]
    /// use rocket::serde::Deserialize;
    ///
    /// #[derive(Deserialize)]
    /// struct Task {
    ///     id: usize,
    ///     complete: bool,
    ///     text: String,
    /// }
    ///
    /// # Client::_test(|_, _, response| {
    /// let response: LocalResponse = response;
    /// let task = response.into_json::<Task>();
    /// # });
    /// ```
    #[cfg(feature = "json")]
    #[cfg_attr(nightly, doc(cfg(feature = "json")))]
    pub $($prefix)? fn into_json<T>(self) -> Option<T>
        where T: Send + serde::de::DeserializeOwned + 'static
    {
        if self._response().body().is_none() {
            return None;
        }

        self._into_json() $(.$suffix)?
    }

    /// Consumes `self` and deserializes its body as MessagePack without
    /// buffering in memory.
    ///
    /// If deserialization fails or the body is unset in the response, returns
    /// `None`. Otherwise, returns `Some`.
    ///
    /// # Example
    ///
    /// ```rust
    #[doc = $doc_prelude]
    /// use rocket::serde::Deserialize;
    ///
    /// #[derive(Deserialize)]
    /// struct Task {
    ///     id: usize,
    ///     complete: bool,
    ///     text: String,
    /// }
    ///
    /// # Client::_test(|_, _, response| {
    /// let response: LocalResponse = response;
    /// let task = response.into_msgpack::<Task>();
    /// # });
    /// ```
    #[cfg(feature = "msgpack")]
    #[cfg_attr(nightly, doc(cfg(feature = "msgpack")))]
    pub $($prefix)? fn into_msgpack<T>(self) -> Option<T>
        where T: Send + serde::de::DeserializeOwned + 'static
    {
        if self._response().body().is_none() {
            return None;
        }

        self._into_msgpack() $(.$suffix)?
    }

    #[cfg(test)]
    #[allow(dead_code)]
    fn _ensure_impls_exist() {
        fn is_debug<T: std::fmt::Debug>() {}
        is_debug::<Self>();
    }
}}
