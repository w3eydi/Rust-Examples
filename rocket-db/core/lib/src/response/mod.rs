//! Types and traits to build and send responses.
//!
//! The return type of a Rocket handler can be any type that implements the
//! [`Responder`](crate::response::Responder) trait, which means that the type knows
//! how to generate a [`Response`]. Among other things, this module contains
//! several such types.
//!
//! # Composing
//!
//! Many of the built-in `Responder` types _chain_ responses: they take in
//! another `Responder` and add, remove, or change information in the response.
//! In other words, many `Responder` types are built to compose well. As a
//! result, you'll often have types of the form `A<B<C>>` consisting of three
//! `Responder`s `A`, `B`, and `C`. This is normal and encouraged as the type
//! names typically illustrate the intended response.

mod responder;
mod redirect;
mod response;
mod debug;
mod body;

pub(crate) mod flash;

pub mod content;
pub mod status;
pub mod stream;

#[doc(hidden)]
pub use rocket_codegen::Responder;

pub use self::response::{Response, Builder};
pub use self::body::Body;
pub use self::responder::Responder;
pub use self::redirect::Redirect;
pub use self::flash::Flash;
pub use self::debug::Debug;

/// Type alias for the `Result` of a [`Responder::respond_to()`] call.
pub type Result<'r> = std::result::Result<Response<'r>, crate::http::Status>;
