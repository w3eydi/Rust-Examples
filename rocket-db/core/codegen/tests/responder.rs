use rocket::local::asynchronous::Client;
use rocket::http::{Status, ContentType, Cookie};
use rocket::response::Responder;
use rocket::serde::json::Json;
use rocket::http::Accept;

#[derive(Responder)]
pub enum Foo<'r> {
    First(String),
    #[response(status = 500)]
    Second(Vec<u8>),
    #[response(status = 404, content_type = "html")]
    Third {
        responder: &'r str,
        ct: rocket::http::ContentType,
    },
    #[response(status = 105)]
    Fourth {
        string: &'r str,
        ct: rocket::http::ContentType,
    },
}

#[rocket::async_test]
async fn responder_foo() {
    let client = Client::debug_with(vec![]).await.expect("valid rocket");
    let local_req = client.get("/");
    let req = local_req.inner();

    let mut r = Foo::First("hello".into())
        .respond_to(req)
        .expect("response okay");

    assert_eq!(r.status(), Status::Ok);
    assert_eq!(r.content_type(), Some(ContentType::Plain));
    assert_eq!(r.body_mut().to_string().await.unwrap(), "hello");

    let mut r = Foo::Second("just a test".into())
        .respond_to(req)
        .expect("response okay");

    assert_eq!(r.status(), Status::InternalServerError);
    assert_eq!(r.content_type(), Some(ContentType::Binary));
    assert_eq!(r.body_mut().to_string().await.unwrap(), "just a test");

    let mut r = Foo::Third { responder: "well, hi", ct: ContentType::JSON }
        .respond_to(req)
        .expect("response okay");

    assert_eq!(r.status(), Status::NotFound);
    assert_eq!(r.content_type(), Some(ContentType::HTML));
    assert_eq!(r.body_mut().to_string().await.unwrap(), "well, hi");

    let mut r = Foo::Fourth { string: "goodbye", ct: ContentType::JSON }
        .respond_to(req)
        .expect("response okay");

    assert_eq!(r.status().code, 105);
    assert_eq!(r.content_type(), Some(ContentType::JSON));
    assert_eq!(r.body_mut().to_string().await.unwrap(), "goodbye");
}

#[derive(Responder)]
#[response(content_type = "plain")]
pub struct Bar<'r> {
    responder: Foo<'r>,
    other: ContentType,
    third: Cookie<'static>,
    #[response(ignore)]
    _yet_another: String,
}

#[rocket::async_test]
async fn responder_bar() {
    let client = Client::debug_with(vec![]).await.expect("valid rocket");
    let local_req = client.get("/");
    let req = local_req.inner();

    let mut r = Bar {
        responder: Foo::Second("foo foo".into()),
        other: ContentType::HTML,
        third: Cookie::new("cookie", "here!"),
        _yet_another: "uh..hi?".into()
    }.respond_to(req).expect("response okay");

    assert_eq!(r.status(), Status::InternalServerError);
    assert_eq!(r.content_type(), Some(ContentType::Plain));
    assert_eq!(r.body_mut().to_string().await.unwrap(), "foo foo");
    assert_eq!(r.headers().get_one("Set-Cookie"), Some("cookie=here!"));
}

#[derive(Responder)]
#[response(content_type = "application/x-custom")]
pub struct Baz {
    responder: &'static str,
}

#[rocket::async_test]
async fn responder_baz() {
    let client = Client::debug_with(vec![]).await.expect("valid rocket");
    let local_req = client.get("/");
    let req = local_req.inner();

    let mut r = Baz { responder: "just a custom" }
        .respond_to(req)
        .expect("response okay");

    assert_eq!(r.status(), Status::Ok);
    assert_eq!(r.content_type(), Some(ContentType::new("application", "x-custom")));
    assert_eq!(r.body_mut().to_string().await.unwrap(), "just a custom");
}

// The bounds `Json<T>: Responder, E: Responder` will be added to the generated
// implementation. This would fail to compile otherwise.
#[derive(Responder)]
enum MyResult<'a, T, E, H1, H2> {
    Ok(Json<T>),
    #[response(status = 404)]
    Err(E, H1, H2),
    #[response(status = 500)]
    Other(&'a str),
}

#[rocket::async_test]
async fn generic_responder() {
    let client = Client::debug_with(vec![]).await.expect("valid rocket");
    let local_req = client.get("/");
    let req = local_req.inner();

    let v: MyResult<_, (), ContentType, Cookie<'static>> = MyResult::Ok(Json("hi"));
    let mut r = v.respond_to(req).unwrap();
    assert_eq!(r.status(), Status::Ok);
    assert_eq!(r.content_type().unwrap(), ContentType::JSON);
    assert_eq!(r.body_mut().to_string().await.unwrap(), "\"hi\"");

    let v: MyResult<(), &[u8], _, _> = MyResult::Err(&[7, 13, 23], ContentType::JPEG, Accept::Text);
    let mut r = v.respond_to(req).unwrap();
    assert_eq!(r.status(), Status::NotFound);
    assert_eq!(r.content_type().unwrap(), ContentType::JPEG);
    assert_eq!(r.body_mut().to_bytes().await.unwrap(), vec![7, 13, 23]);

    let v: MyResult<(), &[u8], ContentType, Accept> = MyResult::Other("beep beep");
    let mut r = v.respond_to(req).unwrap();
    assert_eq!(r.status(), Status::InternalServerError);
    assert_eq!(r.content_type().unwrap(), ContentType::Text);
    assert_eq!(r.body_mut().to_string().await.unwrap(), "beep beep");
}
