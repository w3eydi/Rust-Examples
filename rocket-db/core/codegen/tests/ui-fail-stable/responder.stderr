error: need at least one field
 --> $DIR/responder.rs:4:1
  |
4 | struct Thing1;
  | ^^^^^^

error: [note] error occurred while deriving `Responder`
 --> $DIR/responder.rs:3:10
  |
3 | #[derive(Responder)]
  |          ^^^^^^^^^
  |
  = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: need at least one field
 --> $DIR/responder.rs:7:14
  |
7 | struct Thing2();
  |              ^^

error: [note] error occurred while deriving `Responder`
 --> $DIR/responder.rs:6:10
  |
6 | #[derive(Responder)]
  |          ^^^^^^^^^
  |
  = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: need at least one field
  --> $DIR/responder.rs:13:12
   |
13 | enum Foo { Bark, }
   |            ^^^^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:12:10
   |
12 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: only one lifetime is supported
  --> $DIR/responder.rs:16:14
   |
16 | struct Thing4<'a, 'b>(&'a str, &'b str);
   |              ^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:15:10
   |
15 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid or unknown content type
  --> $DIR/responder.rs:25:27
   |
25 | #[response(content_type = "")]
   |                           ^^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:24:10
   |
24 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid or unknown content type
  --> $DIR/responder.rs:29:27
   |
29 | #[response(content_type = "idk")]
   |                           ^^^^^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:28:10
   |
28 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid value: expected string literal
  --> $DIR/responder.rs:33:27
   |
33 | #[response(content_type = 100)]
   |                           ^^^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:32:10
   |
32 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: status must be in range [100, 599]
  --> $DIR/responder.rs:37:21
   |
37 | #[response(status = 8)]
   |                     ^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:36:10
   |
36 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid value: expected unsigned integer literal
  --> $DIR/responder.rs:41:21
   |
41 | #[response(status = "404")]
   |                     ^^^^^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:40:10
   |
40 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid value: expected unsigned integer literal
  --> $DIR/responder.rs:45:21
   |
45 | #[response(status = "404", content_type = "html")]
   |                     ^^^^^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:44:10
   |
44 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid value: expected string literal
  --> $DIR/responder.rs:49:41
   |
49 | #[response(status = 404, content_type = 120)]
   |                                         ^^^

error: [note] error occurred while deriving `Responder`
  --> $DIR/responder.rs:48:10
   |
48 | #[derive(Responder)]
   |          ^^^^^^^^^
   |
   = note: this error originates in the derive macro `Responder` (in Nightly builds, run with -Z macro-backtrace for more info)
